Initial Implementation

joshmcrae opened the pull request:
This pull request introduces a daemon satisfying the general spec in dokku/dokku#1559. In the interest of simplicity and consistency with other Dokku projects it is implemented in Bash and uses only `netcat` for communication with the socket.

Features:
* Daemon listens on a UNIX domain socket (easily mounted in a Docker container)
* Commands issued by clients match the format of those issued to the `dokku` cli
* Responses are sent as line-delimited JSON
* No authentication layer (local connections only)
* Only one client and one command can be served at a time

One slight deviation from the spec is the response format. Line-delimited JSON makes it easier for clients to consume responses as it doesn't require the use of a streaming parser. In this way both commands and responses are delimited with a newline character and more primitive clients can be built.

Validation is currently only run against command names (sourced from `dokku help`) and not their arguments. A list of valid commands is generated when the daemon starts to speed up responses however this does mean that commands for newly installed plugins will be invalid until the daemon is restarted. Perhaps a 'refresh' signal could be sent over the socket to rebuild this list?

Happy to hear any thoughts on this and make changes as required.

josegonzalez commented:
Fucking beautiful. I'll take a look at this in detail later tonight, but a few questions/thoughts before I do:
- Are you happy with how it came out? Anything at all you would change?
- :+1: on that refresh thing. My initial thought was to implement validation using [docopt](https://github.com/docopt/docopt) - I actually have a new prototype ssh client in go with docopt - so this fits in pretty well if that. We should support both a "refresh" and "non-refresh" mode to ease testing as well.
- Too bad re: client connection limitations. Is this because of the specific implementation language, tooling or some other reason?
- You should document the spec in the readme. Even if we change the spec, it's useful to know the limitations of this daemon.
- How do you think the current implementation would fit into a web-api-wrapper interface? Just curious.
- Did you come up with the current implementation based on any other OS software that does similar?

joshmcrae commented:
@josegonzalez Overall I'm happy with the structure of the daemon but I do have some concerns regarding socket connections. Like you say, it isn't ideal that each client has to close its connection before the next is served. This is a limitation of `netcat`; it simply listens on a socket treating stdout and stdin as read and write streams for the currently connected client. This might lead to a situation where client A writes a command, disconnects and then client B receives the response.

Doing some research this morning I discovered `socat` which seems to be a little more powerful. I'd like to test this out as a replacement for netcat.

docopt looks like an excellent tool for validating help output (assuming the usage text is correct). If the API is already available in a standalone program - even better. My idea for refreshing would be as simple as 

```
< commands-reload
> {"ok":true,"output":"Command validation updated"}
```

and any commands thereafter would be just as fast. As a mode do you mean `dokku help` would be consulted for every command run against the daemon but turned off in production for speed?

In terms of usage in an API I would expect a connection and command be made to the daemon per API request. Of course for an application framework that supported it, it would make more sense for the socket connection to be persistent. Responses from dokku-daemon return the verbatim output of the command so it would be up to the wrapping API to format the data into structure JSON. By this I mean turning rows in arrays, tables into objects etc. The daemon itself is very low level in this regard.

I didn't base this implementation off of any other projects.

I'll update the README and test an implementation using `socat`

josegonzalez commented:
I've enabled travis support btw.

joshmcrae commented:
I've updated dokku-daemon to use `socat` in place of `netcat`. This means that multiple client connections can now be handled simultaneously. Actual interaction with dokku is still protected by a lock so that only one command is executed at a time. I've added these details to the README.

Sorry about the number of commits - I've had zero experience with CI before now and had some trouble getting the tests to run.

josegonzalez commented:
@joshmcrae thoughts on a cache-clearing mechanism?

@michaelshobbs thoughts? If this works, it's mergeable for me.

josegonzalez commented:
One thing I just thought about: how do we want to represent anything where dokku expects stdin or a filepath? Stuff like ssl cert handling comes to mind.

michaelshobbs commented:
aw man! it even has tests!!!! :+1:

josegonzalez commented:
@joshmcrae I'm going to merge this and file two issues regarding stdin/filepaths and cache clearing. 

I've also added you to the dokku org so you can continue working on the dokku-daemon as you see fit (you've already done excellent work!). No pressure of course :)
